File: ruby/huffman_coding/huffman_coding.rb
require '../linked_list/linked_list.rb'

class HuffmanNode < Node
  attr_accessor :symbol, :left, :right, :prefix
  def initialize(symbol, freq)
    super(freq)
    @symbol = symbol
    @prefix = ""
  end
end

class HuffmanTree
  attr_reader :head

  # accepts an array of symbol and frequence pairs
  # ex [ ['a', 0], ['b', 1], ...]
  def initialize(symbol_and_freq_list)
    @head = nil
    @falseNodeChar = '$'
    @ll = LinkedList.new
    @realNodes = build_nodes(symbol_and_freq_list)
    @realNodes.each { |n| @ll.add_node(n) }
    build_tree
    build_prefixes
  end

  def prefix_for(symbol)
    @realNodes.each { |hn|
      return hn.prefix if hn.symbol == symbol
    }
    nil
  end

  def symbol_and_prefix_list
    @realNodes.map { |hn|
      [hn.symbol, hn.prefix]
    }
  end

  private

  def build_nodes(symbol_and_freq_list)
    symbol_and_freq_list.map { |symbol_and_freq|
      symbol = symbol_and_freq[0]
      freq = symbol_and_freq[1]
      HuffmanNode.new(symbol, freq)
    }
  end

  def build_tree
    # continuously pop 2 nodes from the linked list

    while true
      node1 = @ll.pop
      node2 = @ll.pop

      # if there is only one node left, that's the head
      unless node2
        @head = node1
        return
      end

      # create a new node with combined frequency of both nodes
      # the new node is the parent of both nodes
      # node1 is left because it's value would be smaller, node2 is right
      new_freq = node1.value + node2.value
      new_node = HuffmanNode.new(@falseNodeChar, new_freq)
      new_node.left = node1
      new_node.right = node2
      # reinsert (sorted) the combined node
      @ll.add_node(new_node)
    end
  end

  def build_prefixes
    build_prefixes_recurse(@head, nil, false)
  end

  def build_prefixes_recurse(current, parent, went_right)
    return unless current

    if parent
      current.prefix = parent.prefix
      current.prefix += went_right ? "0" : "1"
    end

    build_prefixes_recurse(current.left, current, false)
    build_prefixes_recurse(current.right, current, true)
  end
end

File: javascript/huffman_coding/huffman_coding.js
const HuffmanNode = function (symbol, freq) {
  this.symbol = symbol;
  this.freq = freq;
  this.left = null;
  this.right = null;
  this.next = null;
  this.prefix = '';
};

const SortedLinkedList = function () {
  this.head = null;
};

SortedLinkedList.prototype.insert = function (node) {
  if (!this.head) {
    this.head = node;
    return;
  }

  if (node.freq < this.head.freq) {
    node.next = this.head;
    this.head = node;
    return;
  }

  let current = this.head;
  while (current.next && current.next.freq < node.freq) {
    current = current.next;
  }
  node.next = current.next;
  current.next = node;
};

SortedLinkedList.prototype.pop = function () {
  if (!this.head) {
    return;
  }

  const ret = this.head;
  this.head = this.head.next;
  return ret;
};

SortedLinkedList.prototype.printList = function () {
  let current = this.head;
  process.stdout.write('[ ');
  while (current) {
    process.stdout.write("(" + current.symbol + ", " + current.freq + ") ");
    current = current.next;
  }
  console.log(']');
};

const HuffmanTree = function (symbol_and_freq_list) {
  this.head = null;
  this.combinedNodeChar = '$';
  this.ll = new SortedLinkedList();
  this.nodes = this.buildNodes(symbol_and_freq_list);
  for (let n of this.nodes) {
    this.ll.insert(n);
  }
  this.buildTree();
  this.buildPrefixes(this.head);
};

HuffmanTree.prototype.buildNodes = function (symbol_and_freq_list) {
  const ret = [];
  symbol_and_freq_list.forEach((e) => {
    ret.push(new HuffmanNode(e[0], e[1]));
  });
  return ret;
};

HuffmanTree.prototype.buildTree = function () {
  let building = true;
  while (building) {
    const node1 = this.ll.pop();
    const node2 = this.ll.pop();

    if (!node2) {
      this.head = node1;
      return;
    }

    const newFreq = node1.freq + node2.freq;
    const newNode = new HuffmanNode(this.combinedNodeChar, newFreq);
    newNode.left = node1;
    newNode.right = node2;
    this.ll.insert(newNode);
  }
};

HuffmanTree.prototype.buildPrefixes = function (current, parent, lastWentRight) {
  if (!current) {
    return;
  }

  if (parent) {
    current.prefix = parent.prefix;
    current.prefix += lastWentRight ? "0" : "1";
  }

  this.buildPrefixes(current.left, current, false);
  this.buildPrefixes(current.right, current, true);
};

HuffmanTree.prototype.printTree = function (current, depth) {
  if (!current) { return; }
  const fDepth = depth || 0;
  this.printTree(current.right, fDepth + 1);
  for (let i = 0; i < fDepth; i++) {
    process.stdout.write("---|");
  }
  console.log(current.symbol + ", " + current.freq + ", " + current.prefix);
  this.printTree(current.left, fDepth + 1);
};

HuffmanTree.prototype.symbol_and_prefix_list = function () {
  return this.nodes.map(e => [e.symbol, e.prefix]);
};

const ht = new HuffmanTree([
  ['a', 0],
  ['b', 1],
  ['c', 2],
  ['d', 3],
  ['e', 4]
]);

// console.log(ht.ll);
console.log(ht.symbol_and_prefix_list);
ht.printTree(ht.head);